### PR TITLE
feat: support envoy & ratelimit pod affinity and tolerations.

### LINKED ISSUE #1396: Support Node Affinity and Tolerations for EnvoyProxy Pods
#### *Description*:

I am trying to run a set of proxy on a restricted node pool. The pod needs `affinity` and `tolerations` set to be correctly scheduled.

I would assume it would look something like this:
``` yaml
apiVersion: config.gateway.envoyproxy.io/v1alpha1
kind: EnvoyProxy
metadata:
  name: router-public
  namespace: envoy-gateway-system
spec:
  provider:
    type: Kubernetes
    kubernetes:
      envoyDeployment:
        replicas: 2
        pod:
          affinity:
            nodeAffinity:
              requiredDuringSchedulingIgnoredDuringExecution:
                nodeSelectorTerms:
                  - matchExpressions:
                      - key: cloud.google.com/gke-nodepool
                        operator: In
                        values:
                          - router-node
          tolerations:
            - effect: NoSchedule
              key: node-type
              operator: Exists
              value: "router"
``` 
Where `affinity` and `tolerations`  are propagated to the pod spec in the deployment.

In general, would it be better to allow any fields under `spec.provider.pod` but exclude anything eg controls? i.e. A user can add anything there and the will be merged to the deployment, but certain fields (such as `containers`, `initContainer`) will not be allowed. 

#### *Relevant Links*:
- https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/
- https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/

### PR DESCRIPTION
**Pull request by qicz**
Fixes #1396 

support envoy & ratelimit pod affinity and tolerations.

```yaml
apiVersion: config.gateway.envoyproxy.io/v1alpha1
kind: EnvoyProxy
metadata:
  namespace: envoy-gateway-system
  name: test
spec:
  provider:
    type: Kubernetes
    kubernetes:
      envoyService:
        type: LoadBalancer
      envoyDeployment:
        replicas: 2
        container:
          env:
          - name: env_a
            value: env_a_value
          - name: env_b
            value: env_b_name
          image: "envoyproxy/envoy-dev:latest"
          resources:
            requests:
              cpu: 100m
              memory: 512Mi
          securityContext:
            runAsUser: 2000
            allowPrivilegeEscalation: false
        pod:
          annotations:
            key1: val1
            key2: val2
          affinity:
            nodeAffinity:
              requiredDuringSchedulingIgnoredDuringExecution:
                nodeSelectorTerms:
                - matchExpressions:
                  - key: cloud.google.com/gke-nodepool
                    operator: In
                    values:
                    - router-node
          tolerations:
          - effect: NoSchedule
            key: node-type
            operator: Exists
            value: "router"
          securityContext:
            runAsUser: 1000
            runAsGroup: 3000
            fsGroup: 2000
            fsGroupChangePolicy: "OnRootMismatch"
```

### CODE REVIEW

**Comment by qicz** [comment]
when #1294 merged, will append more unit tests.

**Comment by qicz** [comment]
another choice about #1379 is that we can provide the API `corev1.PodSpec` for `Deployment`with some default settings, such as certs, and volumes. the user can override the default settings.

File: api/config/v1alpha1/shared_types.go [hunk 1]
```diff
@@ -75,6 +75,21 @@ type KubernetesPodSpec struct {
...
+	// +mapType=atomic
+	NodeSelector map[string]string `json:"nodeSelector,omitempty"`
+
+	// If specified, the pod's scheduling constraints.
+	// +optional
+	Affinity *corev1.Affinity `json:"affinity,omitempty" protobuf:"bytes,18,opt,name=affinity"`
+
+	// If specified, the pod's tolerations.
+	// +optional
+	Tolerations []corev1.Toleration `json:"tolerations,omitempty" protobuf:"bytes,22,opt,name=tolerations"`
```
Comment (arkodg) [inline api/config/v1alpha1/shared_types.go:92]: same as above, is the protobuf str needed ?

File: api/config/v1alpha1/shared_types.go [hunk 2]
```diff
@@ -75,6 +75,21 @@ type KubernetesPodSpec struct {
 	//
 	// +optional
 	SecurityContext *corev1.PodSecurityContext `json:"securityContext,omitempty"`
+
+	// NodeSelector is a selector which must be true for the pod to fit on a node.
+	// Selector which must match a node's labels for the pod to be scheduled on that node.
+	// More info: https://kubernetes.io/docs/concepts/configuration/assign-pod-node/
+	// +optional
+	// +mapType=atomic
+	NodeSelector map[string]string `json:"nodeSelector,omitempty"`
```
Comment (arkodg) [inline api/config/v1alpha1/shared_types.go:84]: can we do without `NodeSelector` and rely on `Affinity` ?

Comment (qicz) [reply api/config/v1alpha1/shared_types.go:84]: `corev1.PodSpec` define them as different property, so I think we should keep this.

**Comment by zhaohuabing** [comment]
To ensure we're prepared to support additional features in K8s resource spec - such as HPA, VPA, volumes, volumeMounts and more - I suggest we establish a clear design beforehand.

In my research, I came across how Istio handles Ingress Gateway customization. It's a complex process but could be an excellent reference for our EG implementation. You can find more information on KubernetesResourcesSpec at the following link:

https://istio.io/latest/docs/reference/config/istio.operator.v1alpha1/#KubernetesResourcesSpec

**Review by zhaohuabing — COMMENTED** [review]
To ensure we're prepared to support additional features in K8s resource spec - such as HPA, VPA, volumes, volumeMounts and more - I suggest we establish a clear design beforehand.

In my research, I came across how Istio handles Ingress Gateway customization. Its configuration is quite complex but could be an excellent reference for our EG implementation. You can find more information on KubernetesResourcesSpec at the following link:

https://istio.io/latest/docs/reference/config/istio.operator.v1alpha1/#KubernetesResourcesSpec

File: api/config/v1alpha1/shared_types.go (hunk 2 above)
Comment (arkodg) [reply api/config/v1alpha1/shared_types.go:84]: reframing, which use case does `NodeSelector` enable that `Affinity` cannot do today ?

**Comment by arkodg** [comment]
> To ensure we're prepared to support additional features in K8s resource spec - such as HPA, VPA, volumes, volumeMounts and more - I suggest we establish a clear design beforehand.
> 
> In my research, I came across how Istio handles Ingress Gateway customization. Its configuration is quite complex but could be an excellent reference for our EG implementation. You can find more information on KubernetesResourcesSpec at the following link:
> 
> https://istio.io/latest/docs/reference/config/istio.operator.v1alpha1/#KubernetesResourcesSpec

hey @zhaohuabing if you are referring to [overlays](https://istio.io/latest/docs/reference/config/istio.operator.v1alpha1/#K8sObjectOverlay) in Istio, im a -1 on it because as a end user, if found it extremely hard to generate the right config
regarding additional features/fields we can raise them as individual issues and debate about the pros/cons of adding it. Since we dont have a clear owner for the entire feature, we haven't enforced a design doc, just accepting/enabling k8s fields based on use cases

File: api/config/v1alpha1/shared_types.go (hunk 2 above)
Comment (qicz) [reply api/config/v1alpha1/shared_types.go:84]: xref: https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#nodeselector

they are similar, but should we keep the same for the k8s API?

**Comment by zhaohuabing** [comment]
> arding additional features/fields we can raise them as individual issues and de

I mean the overall structure of the K8s resource： https://istio.io/latest/docs/reference/config/istio.operator.v1alpha1/#KubernetesResourcesSpec

But I agree that we can add new fields when users ask for them. The current structure of envoyDeployment looks good to me.

**Review by zhaohuabing — DISMISSED** [review]
LGTM

File: api/config/v1alpha1/shared_types.go (hunk 2 above)
Comment (arkodg) [reply api/config/v1alpha1/shared_types.go:84]: I would vote to rm `nodeSelector` from this PR, since afaik the use cases can be solved with `affinity` and even the user has not requested this yet

Comment (qicz) [reply api/config/v1alpha1/shared_types.go:84]: cc @AliceProxy @zirain @Xunzhuo

**Comment by qicz** [comment]
ptal @arkodg

File: internal/infrastructure/kubernetes/ratelimit/testdata/deployments/nodeselector.yaml [hunk 3]
```diff
@@ -0,0 +1,83 @@
+apiVersion: apps/v1
+kind: Deployment
+metadata:
```
Comment (arkodg) [inline internal/infrastructure/kubernetes/ratelimit/testdata/deployments/nodeselector.yaml:3]: is this file needed ?

**Review by zhaohuabing — APPROVED** [review]

File: internal/infrastructure/kubernetes/ratelimit/testdata/deployments/nodeselector.yaml (hunk 3 above)
Comment (qicz) [reply internal/infrastructure/kubernetes/ratelimit/testdata/deployments/nodeselector.yaml]: deleted

**Review by arkodg — APPROVED** [review]
LGTM, thanks for adding this feature !
